### PR TITLE
Accepts a string in reduce()

### PR DESCRIPTION
**Pull request by rbtnn**
This is reduce() of https://github.com/vim/vim/pull/9327. 
Please E9998 and E9999 replace latest error numbers.

|               | Blob | String    | List    | Dict   |
|-----------|------|--------|------|------|
|for-in {expr}       | yes  | yes    | yes  |  -   |
| map({expr}, )     | yes  |  yes   | yes  | yes  |
|mapnew({expr}, )  | yes  |  yes     | yes  | yes  |
| filter({expr}, )  | yes  |  yes  | yes  | yes  |
|redulce({expr}, ) | yes  |  -  → **yes**  | yes  |  -   |

And I'm sorry, there was still useless test for negative index in check functions([patch 8.2.3840](https://github.com/vim/vim/commit/ddc80aff575dd60c04c79621a0358cf0abaac53a)).

### CODE REVIEW

**Comment by brammool** [comment]
thanks for working on this.  There is still a leak reported by ASAN, some more work to do...

**Comment by rbtnn** [comment]
@brammool  I think It's ready. Could you check this, please? 
The failure of `GitHub CI / macos (huge) (pull_request)`  does not appear to be caused by this PR.

File: src/testdir/test_listdict.vim [hunk 1]
```diff
@@ -950,13 +960,22 @@ func Test_reduce()
...
 
+  call assert_fails("vim9 reduce(0, (acc, val) => (acc .. val), '')", 'E9998:')
+  call assert_fails("vim9 reduce({}, (acc, val) => (acc .. val), '')", 'E9998:')
+  call assert_fails("vim9 reduce(0.1, (acc, val) => (acc .. val), '')", 'E9998:')
+  call assert_fails("vim9 reduce(function('tr'), (acc, val) => (acc .. val), '')", 'E9998:')
+  call assert_fails("call reduce('', { acc, val -> acc + val }, 1)", 'E9999:')
+  call assert_fails("call reduce('', { acc, val -> acc + val }, {})", 'E9999:')
+  call assert_fails("call reduce('', { acc, val -> acc + val }, 0.1)", 'E9999:')
+  call assert_fails("call reduce('', { acc, val -> acc + val }, function('tr'))", 'E9999:')
+
```
Comment (yegappan) [inline src/testdir/test_listdict.vim:980]: Can you also add a test for using test_null_string() as the first argument to reduce()?

Comment (rbtnn) [reply src/testdir/test_listdict.vim:980]: Thank you for checking. I'll add the tests.

**Comment by rbtnn** [comment]
Thank you for making  the patch.

**Pull request closed without merging**